frontend/apps/reader/modules/readersearch.lua

```diff
@@ -51,33 +60,99 @@ function ReaderSearch:onShowFulltextSearchInput()
                     text = backward_text,
                     callback = function()
                         if self.input_dialog:getInputText() == "" then return end
-                        UIManager:close(self.input_dialog)
-                        self:onShowSearchDialog(self.input_dialog:getInputText(), 1)
+                        self.last_search_text = self.input_dialog:getInputText()
+                        self.use_regex = self.check_button_regex.checked
+                        self.case_insensitive = self.check_button_case.checked
+                        if self.use_regex and self.ui.document:checkRegex(self.input_dialog:getInputText()) ~= 0 then
+                            UIManager:show(InfoMessage:new{ text = _("Error in regular expression!") })
+                        else
+                            UIManager:close(self.input_dialog)
+                            self:onShowSearchDialog(self.input_dialog:getInputText(), 1, self.use_regex, self.case_insensitive)
+                        end
                     end,
                 },
                 {
                     text = forward_text,
                     is_enter_default = true,
                     callback = function()
                         if self.input_dialog:getInputText() == "" then return end
-                        UIManager:close(self.input_dialog)
-                        self:onShowSearchDialog(self.input_dialog:getInputText(), 0)
+                        self.last_search_text = self.input_dialog:getInputText()
+                        self.use_regex = self.check_button_regex.checked
+                        self.case_insensitive = self.check_button_case.checked
+                        if self.use_regex and self.ui.document:checkRegex(self.input_dialog:getInputText()) ~= 0 then
+                            UIManager:show(InfoMessage:new{ text = _("Error in regular expression!") })
+                        else
+                            UIManager:close(self.input_dialog)
+                            self:onShowSearchDialog(self.input_dialog:getInputText(), 0, self.use_regex, self.case_insensitive)
+                        end
                     end,
                 },
             },
         },
     }
+
+   -- checkboxes
+    self.check_button_regex = self.check_button or CheckButton:new{
+        text = _("Regular expression"),
+        face = Font:getFace("smallinfofont"),
+        checked = self.use_regex,
+        callback = function()
+            if not self.check_button_regex.checked then
+                self.check_button_regex:check()
+            else
+                self.check_button_regex:unCheck()
+            end
+            self.input_dialog:onShow()
```

Comment:
Doesn't feel necessary, as if the buttons are shown, their self.input_dialog container is obviously shown? Or is that just to do some redraw (which would still feel odd).

Reply:
If I don't have it in, the checkbuttons don't get checked (redrawn?)

Reply:
Ah, our super-engineered widgets :) Had to look at the code and at openwithdialog.lua to see how that works. Looks like you have to provide:
`parent = self.input_dialog`, to your CheckButtons, and `:check()/:uncheck()` will then do the proper setDirty on their own area (your self.input_dialog:onShow() would redraw the whole larger dialog, which is not necessary).